reana_job_controller/utils.py

```diff
@@ -69,6 +73,118 @@ def initialize_krb5_token(workflow_uuid):
+            delimiter=delimiter,
+            skipinitialspace=skip_initial_space,
+        )
+        for row in csv_reader:
+            yield {
+                key: value if value not in replacements.keys() else replacements[value]
+                for key, value in row.items()
+            }
+
+
+def motley_cue_auth_strategy_factory(hostname):
+    """
+    Paramiko auth strategy factory that provides oauth based ssh token authentication.
+
+    This auth strategy has been developed against the motley cue implementation of
+    oauth based ssh token authentication on the server side.
+
+    :param hostname: hostname of the ssh node
+    :type hostname: str
+    """
+    # Using a factory to avoid a general dependency on libmytoken, paramiko and pyjwt
+    from libmytoken import get_access_token_from_jwt_mytoken
+    from paramiko.auth_strategy import AuthSource
+    from time import time
+    import jwt
+    import requests
+
+    class MotleyCueTokenAuth(AuthSource):
+        def __init__(self):
+            self._access_token = None
+            self._access_token_expires_on = 0
+            self.hostname = hostname
+            self.username = self._get_deployed_username()
+            super().__init__(username=self.username)
+
+        @property
+        def access_token(self):
+            if not (self._access_token and self._is_access_token_valid()):
+                self._refresh_access_token()
+            return self._access_token
+
+        def authenticate(self, transport):
+            return transport.auth_interactive(
+                username=self.username, handler=self.motley_cue_auth_handler
+            )
+
+        def _is_access_token_valid(self):
+            return (
+                self._access_token_expires_on - time() > 100
+            )  # token should be at least valid for 100 s
+
+        def _get_access_token_expiry_date(self):
+            decoded_token = jwt.decode(
+                self._access_token,
+                options={"verify_signature": False, "verify_aud": False},
+            )
+            return decoded_token["exp"]
+
+        def _get_deployed_username(self):
+            headers = {"Authorization": f"Bearer {self.access_token}"}
+            req = requests.get(
+                f"https://{self.hostname}/user/deploy", headers=headers, verify=True
+            )
+            return req.json()["credentials"]["ssh_user"]
+
+        def motley_cue_auth_handler(self, title, instructions, prompt_list):
+            return [
+                self.access_token if (echo and "Access Token" in prompt) else ""
+                for prompt, echo in prompt_list
+            ]
+
+        def _refresh_access_token(self):
+            self._access_token = get_access_token_from_jwt_mytoken(
+                os.environ.get("HELMHOLTZ_TOP")
+            )
+            self._access_token_expires_on = self._get_access_token_expiry_date()
+
+    return MotleyCueTokenAuth()
+
+
 @singleton
 class SSHClient:
```

Comment:
Having a look at this, given that this is a singleton, I think that the current implementation of SlurmCERN and C4PCompute cannot be used at the same time for mixed workflows, i.e. workflows where jobs run on different backends.

If this is ok we can keep it like this, otherwise a refactor is needed. Of course the refactor would be on us.

@tiborsimko what do you think about this?

Reply:
Opened an issue for this: https://github.com/reanahub/reana-job-controller/issues/456

This means that Slurm@CERN cannot be enabled at the same time as C4P. Does this work for you @giffels ?